Fixed #35515 -- Added auto-importing to shell command. (#18158)
django/django: docs/howto/custom-shell.txt

============================================= 

> How to customize the :djadmin:`shell` command 

Reviewer: does a link in a title work well? I would guess not. Did you build the docs locally?

Regardless, let’s use plain text in the title:

```suggestion
How to customize the ``shell`` command 
```

The link can live only in the body text below.

====================================== 
How to customize the ``shell`` command 
======================================

> To customize the :djadmin:`shell`, create a new :doc:`management command
> </howto/custom-management-commands>` which subclasses 
> ``django.core.management.commands.shell.Command`` and overrides the existing
> ``shell`` management command. See :ref:`Overriding commands 
> <overriding-commands>` for more details. 

Reviewer 2: Maybe
```suggestion
Create a new :doc:`custom management command</howto/custom-management-commands>`
which subclasses the ``django.core.management.commands.shell.Command``,
overriding the existing ``shell`` management command. See the guide on
:ref:`overriding commands <overriding-commands>` for more details.

.. _customizing-shell-auto-imports:
```

Customizing the auto-imports
============================

.. versionadded:: 5.2

> To customize the auto-importing behavior of :djadmin:`shell`, you can override
> the ``get_namespace()`` method. For example, assuming you have a ``polls``
> application in :setting:`INSTALLED_APPS`:

Reviewer 2: Now thinking the "assuming you have a ``polls`` application in :setting:`INSTALLED_APPS`" is distracting

```suggestion
To customize the auto-importing behavior of the :djadmin:`shell` management
command, you can override the ``get_namespace()`` method.

For example:
```

.. code-block:: python
    :caption: ``polls/management/commands/shell.py``

    from django.core.management.commands import shell

    class Command(shell.Command):
        def get_namespace(self):
            from django.urls.base import resolve, reverse

            return {
                **super().get_namespace(),
                "resolve": resolve,
                "reverse": reverse,
            }

> This way, your customized shell will import models from your apps along with the
> ``resolve`` and ``reverse`` methods, allowing you to use them without needing
> additional imports.
> 
> .. note::
> 
>     If you prefer your shell without any auto-imported model, you can omit
>     calling the ``**super().get_namespace()`` method when returning the
>     namespace, as it can be empty.
> 
> 

Reviewer 2: I think it might be better just to have a second example
```suggestion
This way, your custom shell command will import models from your apps along with
the ``resolve`` and ``reverse`` methods. You can then use them without needing
additional imports.

If you prefer not to have any auto-imported models. You can omit calling
``**super().get_namespace()`` in the ``get_namespace()`` method:

.. code-block:: python
    :caption: ``polls/management/commands/shell.py``

    from django.core.management.commands import shell

    class Command(shell.Command):
        def get_namespace(self):
            return {}

```

====================================== 
How to customize the ``shell`` command 
======================================

> First, create a new custom management command which subclasses

Reviewer 3: I feel this needs a small intro paragraph so I'm pushing a small adding.

======================================
How to customize the ``shell`` command
======================================

First, create a new custom management command which subclasses
``django.core.management.commands.shell.Command``, overriding the existing
``shell`` management command. See the guide on :ref:`overriding commands
<overriding-commands>` for more details.

.. _customizing-shell-auto-imports:

Customizing the auto-imports
============================

.. versionadded:: 5.2

> To customize the auto-importing behavior of the :djadmin:`shell` management
> command, you can override its ``get_namespace()`` method.
> 
> For example:

Reviewer: ```suggestion
To customize the automatic import behavior of the :djadmin:`shell` management
command, override its ``get_namespace()`` method. For example:
```

Customizing the auto-imports
============================

.. versionadded:: 5.2

To customize the auto-importing behavior of the :djadmin:`shell` management
command, you can override the ``get_namespace()`` method.

For example:

.. code-block:: python
    :caption: ``polls/management/commands/shell.py``

    from django.core.management.commands import shell

    class Command(shell.Command):
        def get_namespace(self):
            from django.urls.base import resolve, reverse

            return {
                **super().get_namespace(),
                "resolve": resolve,
                "reverse": reverse,
            }

This way, your custom shell command will import models from your apps along with
the ``resolve`` and ``reverse`` methods. You can then use them without needing
additional imports.

If you prefer not to have any auto-imported models. You can omit calling
``**super().get_namespace()`` in the ``get_namespace()`` method:

.. code-block:: python
    :caption: ``polls/management/commands/shell.py``

    from django.core.management.commands import shell

    class Command(shell.Command):
        def get_namespace(self):
            return {}

> 

Reviewer 2: ```suggestion
```

</howto/custom-management-commands>` on how to override an existing command. 

Customizing the shell
=====================

To override the current shell's functionality, you have to override the
``shell.Command.get_namespace`` in order to return a namespace that maps every
name to the object imported. For example, if you want to import ``resolve`` and
``reverse`` methods in you shell, you can edit
``polls/management/commands/shell.py`` to look like this::

    from django.core.management.commands import shell

    class Command(shell.Command):
        def get_namespace(self):
            from django.urls.base import resolve, reverse

            return {
                **super().get_namespace(),
                "resolve": resolve,
                "reverse": reverse,
            }

This way, your customized shell will import models from your apps along with the
``resolve`` and ``reverse`` methods, allowing you to use them without needing
additional imports.

.. note::

    If you prefer your shell without any auto-imported model, you can omit
    calling the ``**super().get_namespace()`` method when returning the
    namespace, as it can be empty.

> Implement your own python shell runner

Reviewer 4: ```suggestion
Implementing your own python shell runner
```

============================================= 
How to customize the ``shell`` command 
=============================================

> Django provides functionality to auto-import project models in the shell. In
> your project, you might want to either avoid importing models or include
> additional imports in the shell. You can do this by subclassing the
> :djadmin:`shell` command and overriding a method. To set up the directory of
> your project you can follow this :doc:`guide
> </howto/custom-management-commands>` on how to override an existing command. 

Reviewer 2: ```suggestion
To customize the :djadmin:`shell`, create a new :doc:`management command
</howto/custom-management-commands>` which subclasses 
``django.core.management.commands.shell.Command`` and overrides the existing
``shell`` management command. See <!INSERT NEW LINK TO "Overriding commands" HERE!>
for more details.
```

    models you want to access.

Customizing the shell
=====================

To override the current shell's functionality, you have to override the
``shell.Command.get_namespace`` in order to return a namespace that maps every
name to the object imported. For example, if you want to import ``resolve`` and
``reverse`` methods in you shell, you can edit
``polls/management/commands/shell.py`` to look like this::

    from django.core.management.commands import shell

    class Command(shell.Command):
        def get_namespace(self):
            from django.urls.base import resolve, reverse

            return {
                **super().get_namespace(),
                "resolve": resolve,
                "reverse": reverse,
            }

This way, your customized shell will import models from your apps along with the
``resolve`` and ``reverse`` methods, allowing you to use them without needing
additional imports.

.. note::

    If you prefer your shell without any auto-imported model, you can omit
    calling the ``**super().get_namespace()`` method when returning the
    namespace, as it can be empty.

> Implementing your own python shell runner

Reviewer 2: I think this is additional in the how to and not related to the auto-importing feature - I would pull this section out to a separate PR
